### PR TITLE
remove empty regions from signals when applying  LabelMasker.transform

### PR DESCRIPTION
**Pull request by mtorabi59**
Addresses #3085.

Changes proposed in this pull request:
This PR is to address the issue of empty `signals` dimensions after applying `labelmasker.transform`. So, basically we first apply the additional mask and then find the list of all existing labels, instead of first finding them and then setting them to zero. This way they will be treated as background from the beginning. As a result, no dimension will be dedicated to those empty regions in `signals` output anymore.
I've also added a test function designated to test the NiftiLabelMasker module.

### CODE REVIEW

**Comment by Remi-Gau** [comment]
@mtorabi59 given how much change there is in some folders because of the reformatting, watch out for merge conflicts theses days !

I will let solve this one, but ping me if you need help.

**Comment by ymzayek** [comment]
Hi @mtorabi59 thanks for working on this! A couple of tips: 

Tests for modules are usually contained within that module under a `tests` directory. Since `signal_extraction.py` is in `nilearn/regions` you will find the relevant test file in `nilearn/regions/tests/test_signal_extraction.py`. So the first thing when changing code in a file is to look to see if any of the already written tests will fail. Indeed this is the case because when I run pytest on that test file I see that `test_signal_extraction_with_maps_and_labels` will fail. So what you could do first is see how you can modify the failing test to test the new behavior and get it to pass.

It may be useful sometimes to add tests at higher levels which I think is why you are testing `NiftiLabelsMasker`. However, this is in the wrong location. Such tests should go in `nilearn/maskers/tests/test_nifti_labels_masker.py`. I know this may be confused with the `masking` module so just to clarify `nilearn/masking.py` and `nilearn/maskers/` are two different modules. So first look at the tests in `test_nifti_labels_masker.py` to see what could be modified or added to test the new behavior (but first run pytest on this test file to check if anything already fails)

Let me know if you have any questions

**Comment by Remi-Gau** [comment]
There is actually other tests in `nilearn/tests/test_masking.py` that probably should go `nilearn/maskers/tests/test_nifti_labels_masker.py`

See for example: 

https://github.com/nilearn/nilearn/blob/32835204227f65204b911fb2fd9ab029968a2182/nilearn/tests/test_masking.py#L552

I am trying to put them back where they belong to avoid confusion.

File: nilearn/tests/test_masking.py [hunk 1]
```diff
@@ -482,6 +482,99 @@ def test_intersect_masks():
...
+
+    # apply atlas_img on img
+    masker = NiftiLabelsMasker(atlas_img,
+                           labels=labels,
+                           standardize=False
+                           )
+    masker.fit()
+    signals = masker.transform(img)
+    expected_signal = np.tile(np.array([[1, 2, 3]]), (10, 1))
+    assert_array_equal(expected_signal, signals)
```
Comment (Remi-Gau) [inline nilearn/tests/test_masking.py:546]:
```suggestion
    assert_array_equal(signals, expected_signal)
```

I have seen both in the nilearn test suite, but I generally find

```python
assert results == expected_result
```

more readable than 

```python
assert expected_result == results
```

For the same reason that we more often write

```python
if length >= 10:
```

than this

```python
if 10 <= length:
```

---

The rule from: ["the art of readable code"](https://mcusoft.files.wordpress.com/2015/04/the-art-of-readable-code.pdf)

### Left-hand side	
The expression “being interrogated,” whose value is more in flux.

### Right-hand side
The expression being compared against, whose value is more constant.

Comment (mtorabi59) [reply nilearn/tests/test_masking.py]: Thanks @Remi-Gau  for explaining this !

**Comment by mtorabi59** [comment]
So I realized that the `img_to_signals_maps` also has the same behavior of returning empty regions and that is why the test fails when comparing the results of `img_to_signals_maps` and `signals_to_img_labels`. I have to solve the issue for `img_to_signals_maps` too so that they give consistent outputs. Should I do change both in the same PR or I should change  `signals_to_img_labels` and comment that test for now until the next PR that I change `img_to_signals_maps`?

**Comment by mtorabi59** [comment]
Also, should I put a DeprecationWarning for the change? e.g. "Starting in version x, the regions removed by mask_img will be removed from region_signals instead of being filled by zeros."

**Comment by ymzayek** [comment]
@mtorabi59 I think in general it's better to keep PRs open and not overwrite the code changes if there is an issue so that we can better discuss the next steps with context. Sometimes things like that come up and we could either decide to handle them in the same PR or open another one. In any case let's take a closer look regarding the consistency of  behavior of those two functions. And indeed if the output changes in public functions we would need to go through a deprecation cycle.

**Comment by mtorabi59** [comment]
@ymzayek yes you are right. I didn't know that syncing my fork will close the PR automatically. It wasn't intended. Yes I had a closer look. Should we discuss that in [the issue #3085](https://github.com/nilearn/nilearn/issues/3085)? or in office hours?

**Comment by ymzayek** [comment]
@mtorabi59 ok yes makes sense what happened, now worries! Yes discussing it either in the issue or office hours (or both) is fine.

**Pull request closed without merging**